File: src/test/java/org/gradle/plugins/fsm/FSMPluginExtensionTest.java
package org.gradle.plugins.fsm;

import org.gradle.api.Project;
import org.gradle.api.artifacts.ProjectDependency;
import org.gradle.api.plugins.JavaPlugin;
import org.gradle.testfixtures.ProjectBuilder;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Collections;
import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;

class FSMPluginExtensionTest {

    private FSMPluginExtension testling;
    private Project project;

    @BeforeEach
    void setUp() {
        project = ProjectBuilder.builder().build();
        project.getPlugins().apply(FSMPlugin.NAME);
        testling = new FSMPluginExtension(project);
    }

    @Test
    void testWebApps() {
        final Project webAppSubprojectA = ProjectBuilder.builder().withParent(project).withName("web_a").build();
        final Project webAppSubprojectB = ProjectBuilder.builder().withParent(project).withName("web_b").build();

        testling.webAppComponent("WebAppA", webAppSubprojectA);
        testling.webAppComponent(webAppSubprojectB);

        assertThat(testling.getWebApps()).containsEntry("WebAppA", webAppSubprojectA);
        assertThat(testling.getWebApps()).containsEntry("web_b", webAppSubprojectB);

        // Ensure the project has a compile dependency on the subprojects
        final Stream<Project> dependencyProjects = project.getConfigurations().getByName(JavaPlugin.IMPLEMENTATION_CONFIGURATION_NAME).getDependencies().stream()
                .filter(dependency -> dependency instanceof ProjectDependency)
                .map(dependency -> ((ProjectDependency) dependency).getDependencyProject());
        assertThat(dependencyProjects).containsExactly(webAppSubprojectA, webAppSubprojectB);
    }

    @Test
    void testMinVersionIsDefault() {
        assertThat(testling.getAppendDefaultMinVersion()).isTrue();
    }


    @Test
    void testCanSetFsmDependencies() {
        final String fsmModuleName = "fsmModuleName";
        testling.setFsmDependencies(Collections.singletonList(fsmModuleName));
        assertThat(testling.getFsmDependencies()).contains(fsmModuleName);
    }


}

File: src/test/kotlin/org/gradle/plugins/fsm/FSMPluginExtensionTest.kt
package org.gradle.plugins.fsm

import org.assertj.core.api.Assertions.assertThat
import org.gradle.api.Project
import org.gradle.api.artifacts.ProjectDependency
import org.gradle.testfixtures.ProjectBuilder
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class FSMPluginExtensionTest {

    private lateinit var testling: FSMPluginExtension
    private lateinit var project: Project

    @BeforeEach
    fun setUp() {
        project = ProjectBuilder.builder().build()
        project.plugins.apply(FSMPlugin.NAME)
        testling = FSMPluginExtension(project)
    }

    @Test
    fun testWebApps() {
        val webAppSubprojectA = ProjectBuilder.builder().withParent(project).withName("web_a").build()
        val webAppSubprojectB = ProjectBuilder.builder().withParent(project).withName("web_b").build()
        testling.webAppComponent("WebAppA", webAppSubprojectA)
        testling.webAppComponent(webAppSubprojectB)
        assertThat(testling.getWebApps()).containsEntry("WebAppA", webAppSubprojectA)
        assertThat(testling.getWebApps()).containsEntry("web_b", webAppSubprojectB)

        // Ensure the project has a compile dependency on the subprojects
        val dependencyProjects = project.configurations.getByName(FSMPlugin.WEBAPPS_CONFIGURATION_NAME)
            .dependencies.filterIsInstance<ProjectDependency>().map { it.dependencyProject(project) }
        assertThat(dependencyProjects).containsExactly(webAppSubprojectA, webAppSubprojectB)
    }

    @Test
    fun testMinVersionIsDefault() {
        assertThat(testling.appendDefaultMinVersion).isTrue
    }


    @Test
    fun testCanSetFsmDependencies() {
        val fsmModuleName = "fsmModuleName"
        testling.fsmDependencies = listOf(fsmModuleName)
        assertThat(testling.fsmDependencies).contains(fsmModuleName)
    }
}